### PR TITLE
feat: custom transformer with wasm

### PR DESCRIPTION
**Pull request by benny-n**
This is a draft of implementing #26 
 
I have written some boilerplate code which will eventually lead to the full implementation of custom transformer with Web Assembly (wasm) and I also included one simple test.

Before I continue with the implementation, there are a few points that should be addressed:
1. where should we receive the input for this transformer (the wasm bytes)? as a special new argument flag? or from replibyte.yaml file?
2. wasm only supports 32/64 bits numbers, so maybe we should add a few more values to `Column`:
    ```rust
    pub enum Column {
         // ...
         NumberValuei32(String, i32),
         NumberValuei64(String, i64),
         NumberValuei128(String, i128),  // not compatible with wasm, unless we truncate
        // ...
    }
    ```
3. the first try of compilation yielded the following error:
**`*mut Ctx` cannot be shared between threads safely
within `wasmer_runtime_core::instance::InstanceInner`, the trait `std::marker::Sync` is not implemented for `*mut Ctx `**

    This is due to fact that `Transformer` is `Sync`.
    as a temporary solution, I removed `Sync` from `Transformer`, of course I realize that this is not a long term solution but I wanted to make the compilation and example test work.
we need to think about how we are going to address this issue.

There will probably be more points to think about in the future, but for now these are the main things that bothered me.
@evoxmusic what do you think?

### CODE REVIEW

File: replibyte/src/transformer/mod.rs [hunk 1]
```diff
@@ -31,7 +32,7 @@ pub fn transformers() -> Vec<Box<dyn Transformer>> {
 }
 
 /// Trait to implement to create a custom Transformer.
-pub trait Transformer: Sync {
+pub trait Transformer {
```
Comment (benny-n) [inline replibyte/src/transformer/mod.rs:35]: temporary solution until we figure out how to deal with this.

**Comment by evoxmusic** [comment]
Thank you @benny-n for starting digging into this feature. I respond inline:

> 1. where should we receive the input for this transformer (the wasm bytes)? as a special new argument flag? or from replibyte.yaml file?

I'd recommend using the `.yaml` file. We can have multiple wasm transformers for different tables. Which can lead to a very long command line. 

> 2. wasm only supports 32/64 bits numbers, so maybe we should add a few more values to Column:

From what I understand, the wasm compiler in any language is able to even translate String into an array of bytes. [Here is an example](https://www.secondstate.io/articles/getting-started-with-rust-function/).

```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn say(s: String) -> String {
  let r = String::from("hello ");
  return r + &s;
}
```

another more complex example using serde

```rust
#[wasm_bindgen]
pub fn solve(params: &str) -> String {
  let ps: (f32, f32, f32) = serde_json::from_str(&params).unwrap();
  let discriminant: f32 = (ps.1 * ps.1) - (4. * ps.0 * ps.2);
  let mut solution: (f32, f32) = (0., 0.);
  if discriminant >= 0. {
    solution.0 = (((-1.) * ps.1) + discriminant.sqrt()) / (2. * ps.0);
    solution.1 = (((-1.) * ps.1) - discriminant.sqrt()) / (2. * ps.0);
    return serde_json::to_string(&solution).unwrap();
  } else {
    return String::from("not real numbers");
  }
}
```

[Here is the list of wasm compilers available](https://webassembly.org/getting-started/developers-guide/)

> 3. concerning Sync

We can see later on, it's a "minor" point even if we need to do some refacto. 

----

The other aspect for me is, how easy it is for a user to create a wasm module? When I look at the [compilers list available](https://webassembly.org/getting-started/developers-guide/), I see that there is no high-level language like Javascript or Typescript. WASM is super performant but requires an advanced setup to get started. Maybe we'll also need to provide intermediate solutions to exec external commands as a transformer?

Example:

```yaml
...
transformer_name: command
transformer_options:
  exec: python -c main.py
```

where the main.py is

```python
import sys
import json

if __name__ == "__main__":
  # get json string from arg 1
  input_str = sys.argv[1]

  # transform string into python dict
  input_json = json.loads(input_str)

  # modify json string
  column_value = input_json["value"]

  # replace the value by ******** 
  modified_json = {"value": len(column_value) * "*"}

  modified_json_str = json.dumps(modified_json)
  print(modified_json) # output json 
```

What do you think @benny-n @fabriceclementz ?

**Comment by benny-n** [comment]
> From what I understand, the wasm compiler in any language is able to even translate String into an array of bytes. [Here is an example](https://www.secondstate.io/articles/getting-started-with-rust-function/).

Yes, I'm aware that we can also pass strings and more complicated data structures to wasm, however passing i32 / i64 to wasm is extremely simple and doesn't even require a use of the `#[wasm_bindgen]` macro, therefore I think we should think about extending the `Column` values to support native i64/i32 numbers anyway. Besides, another advantage of this extension would be to support more `Bson` types (which are different and more versatile than SQL types) in the mongodb source .

> The other aspect for me is, how easy it is for a user to create a wasm module? When I look at the [compilers list available](https://webassembly.org/getting-started/developers-guide/), I see that there is no high-level language like Javascript or Typescript. WASM is super performant but requires an advanced setup to get started.

Something that is also important to notice is that using `wasm` crates significantly enlarges RepliByte's dependency list (+1300~ lines in Cargo.lock) and will lead to longer build times. We should consider these implications on the crate size and dependency list, especially if this feature is not going to be popular among users due to advanced setup.

> Maybe we'll also need to provide intermediate solutions to exec external commands as a transformer?

I thought we didn't want to support adding external commands as transformers due to fear of hurting performance?
Imagine the user has 10+ different command transformers, that would mean 10+ different sub-processes, one for each command.. I can see this heavily impacting RepliByte's performance.
Maybe we could explain this in the `README.md` and make this the user's responsibility to worry about not hurting performance too much with cmd transformers, since after all in this case he is the one that provides the transformers.

@evoxmusic Let me know how would you like me to proceed with the implementation.

**Comment by fabriceclementz** [comment]
I think at first providing a way to execute simple commands would be more convenient for developers.

eg: 
```yaml
transformer_name: command
transformer_options:
  exec: python -c main.py
```

It leaves the choice to develop in your own language (bash, go, python, node ...) and start small, a simple explanation in the README would be enough as mentioned by @benny-n.
By adding more transformers with options, we could decrease the need for custom plugins.

That said, wasm would be a nice feature for edge case with a high quantity of data and but actually, I don't know enough about wasm to be relevant.

**Comment by evoxmusic** [comment]
> Yes, I'm aware that we can also pass strings and more complicated data structures to wasm, however passing i32 / i64 to wasm is extremely simple and doesn't even require a use of the #[wasm_bindgen] macro, therefore I think we should think about extending the Column values to support native i64/i32 numbers anyway. Besides, another advantage of this extension would be to support more Bson types (which are different and more versatile than SQL types) in the mongodb source .

Ok, make sense to me 👍🏽 

> Something that is also important to notice is that using wasm crates significantly enlarges RepliByte's dependency list (+1200~ lines in Cargo.lock) and will lead to longer build times. We should consider these implications on the crate size and dependency list, especially if this feature is not going to be popular among users due to the advanced setup.

Maybe we can create 2 build versions of RepliByte? Or we can keep it simple at the beginning and provide the wasm custom transformer and if not used, then we can remove it. What do you think?

> I thought we didn't want to support adding external commands as transformers due to fear of hurting performance? Imagine the user has 10+ different command transformers, that would mean 10+ different sub-processes, one for each command.. I can see this heavily impacting RepliByte's performance.
Maybe we could explain this in the README.md and make this the user's responsibility to worry about not hurting performance too much with cmd transformers, since after all in this case he is the one that provides the transformers.

I think we can warn the user that it can drastically affect the performance of RepliByte and impact his CPU. If we are clear enough I guess it's acceptable and a good trade-off between wasm or creating a PR for a custom transformer.

**Comment by evoxmusic** [comment]
> By adding more transformers with options, we could decrease the need for custom plugins.

@fabriceclementz I agree, not sure we can cover all the needs, but over time we'll probably cover 98% of them.

**Comment by benny-n** [comment]
> Maybe we can create 2 build versions of RepliByte? Or we can keep it simple at the beginning and provide the wasm custom transformer and if not used, then we can remove it. What do you think?

I think this is a good case for the [cargo 'features' option](https://doc.rust-lang.org/cargo/reference/features.html).
Other developers or users that depend on the crate can disable the 'wasm' feature if they're not interested in it.

In `Cargo.toml` we would have:
```toml
# ...
wasmer = {version = "2.2", optional = true }
wasmer-runtime = {version = "0.16", optional = true }

[features]
wasm = [ "wasmer", "wasmer-runtime" ]
# ...
```

And in the code:
```rust
// ...
#[cfg(feature = "wasm")]
pub mod custom_wasm;
// ...
```

**Comment by benny-n** [comment]
So after some research I decided that the best way to integrate wasm into RepliByte was to use a combination of `wasmer` runtime with [WASI](https://wasi.dev/), which is "is a modular system interface for WebAssembly", that is currently in the [process of being standardized](https://github.com/WebAssembly/WASI).

This design choice immediately made things a lot easier, giving me the option to write clear and simple Rust functions, compile them to `.wasm` binaries for a `wasi` target, and then run them with `wasmer`.

Comfortably enough using `wasi` also solved the typing issue I was struggling with, since with `wasi`'s std implementation we can just pass values to `wasm` modules through stdin.

I created a pretty detailed example in `examples/wasm` with a README file on how to create a custom wasm transformer with Rust, and use it in RepliByte.

I've also tested this feature with the example's transformer, looks like it works just fine 🙂

@evoxmusic From my persperctive this PR is ready for review, let me know what you think.

**Comment by evoxmusic** [comment]
I didn't have enough emojis to show my joy! Let me take a look in a couple of hours 👍🏽 super nice job @benny-n 🚀

**Review by evoxmusic — APPROVED** [review]

File: replibyte/src/transformer/custom_wasm.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            // Write to the stdin pipe
+            writeln!(wasi_stdin, "{}", value)?;
+        }
+
+        // Call the `_start` function
+        let start = instance.exports.get_function("_start")?;
+        start.call(&[])?; //TODO support calling with parameters
+
+        let mut state = self.wasm_config.wasi_env.state();
+        let wasi_stdout = state.fs.stdout_mut()?.as_mut().unwrap();
```
[the remark this answers is not in the document]
Comment (benny-n) [reply replibyte/src/transformer/custom_wasm.rs]: refactored 🙂

**Pull request merged**